### ISSUE TITLE
linked service cannot be deleted

### ISSUE THREAD

**Issue by yangcao77** [openshift/odo#4750]
/kind bug

## What versions of software are you using?

**Operating System:**
macos + minikube

**Output of `odo version`:**
```
$ odo version
odo v2.2.1 (17a078b67)
```

## How did you run odo exactly?
```
odo create service ...
odo push
odo delete service
odo push
...

odo delete
```

## Actual behavior
the linked service persists, even the component has been deleted on cluster

## Expected behavior
the linked service should be removed from cluster

## Any logs, error output, etc?
```
$ odo push

Validation
 ✓  Validating the devfile [37538ns]

Creating Kubernetes resources for component nodejs-node-test-wzrw
Created service "ServiceBinding/example-servicebinding" on the cluster; refer "odo link -h" to know how to link it to the component
 ✓  Waiting for component to start [3s]
 ✓  Waiting for component to start [8ms]
 ⚠  Unable to create ingress, missing host information for Endpoint http-3000, please check instructions on URL creation (refer `odo url create --help`)


Applying URL changes
W0525 11:03:37.148420   88283 warnings.go:70] extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
 ✓  URLs are synced with the cluster, no changes are required.

Syncing to component nodejs-node-test-wzrw
 ✓  Checking files for pushing [2ms]
 ✓  Syncing files to the component [287ms]

Executing devfile commands for component nodejs-node-test-wzrw
 ✓  Waiting for component to start [1ms]
 ✓  Executing install command "npm install" [9s]
 ✓  Executing run command "npm start" [1s]

Pushing devfile component "nodejs-node-test-wzrw"
 ✓  Changes successfully pushed to component

$ odo service list
NAME                                      AGE
ServiceBinding/example-servicebinding     1m6s

$ odo delete
? Are you sure you want to delete the devfile component: nodejs-node-test-wzrw? Yes

Gathering information for component nodejs-node-test-wzrw
 ✓  Checking status for component [7ms]

Deleting component nodejs-node-test-wzrw
 ✓  Deleting Kubernetes resources for component [5ms]
 ✓  Successfully deleted component
$ kubectl get po
NAME                                     READY   STATUS        RESTARTS   AGE
nodejs-node-test-wzrw-8666bd99f5-nlwmz   0/1     Terminating   0          107s
$ kubectl get po
No resources found in default namespace.
$ kubectl get ServiceBinding/example-servicebinding -o yaml
apiVersion: binding.operators.coreos.com/v1alpha1
kind: ServiceBinding
metadata:
  creationTimestamp: "2021-05-25T15:03:34Z"
  finalizers:
  - finalizer.servicebinding.openshift.io
  generation: 1
  labels:
    app: app
    app.kubernetes.io/instance: nodejs-node-test-wzrw
    app.kubernetes.io/managed-by: odo
    app.kubernetes.io/managed-by-version: v2.2.1
    app.kubernetes.io/part-of: app
  managedFields:
  - apiVersion: binding.operators.coreos.com/v1alpha1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:finalizers:
          .: {}
          v:"finalizer.servicebinding.openshift.io": {}
      f:status:
        .: {}
        f:conditions:
          .: {}
          k:{"type":"CollectionReady"}:
            .: {}
            f:lastTransitionTime: {}
            f:message: {}
            f:reason: {}
            f:status: {}
            f:type: {}
          k:{"type":"Ready"}:
            .: {}
            f:lastTransitionTime: {}
            f:message: {}
            f:reason: {}
            f:status: {}
            f:type: {}
        f:secret: {}
    manager: manager
    operation: Update
    time: "2021-05-25T15:03:34Z"
  - apiVersion: binding.operators.coreos.com/v1alpha1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:labels:
          .: {}
          f:app: {}
          f:app.kubernetes.io/instance: {}
          f:app.kubernetes.io/managed-by: {}
          f:app.kubernetes.io/managed-by-version: {}
          f:app.kubernetes.io/part-of: {}
      f:spec:
        .: {}
        f:application:
          .: {}
          f:group: {}
          f:name: {}
          f:resource: {}
          f:version: {}
        f:bindAsFiles: {}
        f:services: {}
    manager: odo
    operation: Update
    time: "2021-05-25T15:03:34Z"
  name: example-servicebinding
  namespace: default
  resourceVersion: "81524"
  uid: 1153195f-cc62-47dd-9671-31e1ba7e5622
spec:
  application:
    group: apps
    name: nodejs-rest-http-crud
    resource: deployments
    version: v1
  bindAsFiles: true
  services:
  - group: postgresql.example.dev
    kind: Database
    name: pg-instance
    version: v1alpha1
status:
  conditions:
  - lastTransitionTime: "2021-05-25T15:03:34Z"
    message: no matches for kind "Database" in version "postgresql.example.dev/v1alpha1"
    reason: ErrorReadingServices
    status: "False"
    type: CollectionReady
  - lastTransitionTime: "2021-05-25T15:03:34Z"
    message: no matches for kind "Database" in version "postgresql.example.dev/v1alpha1"
    reason: ProcessingError
    status: "False"
    type: Ready
  secret: ""
$ odo service list
NAME                                      AGE
ServiceBinding/example-servicebinding     3m25s
$ odo service list
NAME                                      AGE
ServiceBinding/example-servicebinding     12m55s

$ odo service delete ServiceBinding/example-servicebinding
? Are you sure you want to delete ServiceBinding/example-servicebinding Yes
 ✓  Waiting for service to be deleted [14ms]
Service "ServiceBinding/example-servicebinding" has been successfully deleted
$ odo service list
NAME                                      AGE
ServiceBinding/example-servicebinding     13m26s
$ cat devfile.yaml 
commands:
- exec:
    commandLine: npm install
    component: runtime
    group:
      isDefault: true
      kind: build
    workingDir: /project
  id: install
- exec:
    commandLine: npm start
    component: runtime
    group:
      isDefault: true
      kind: run
    workingDir: /project
  id: run
- exec:
    commandLine: npm run debug
    component: runtime
    group:
      isDefault: true
      kind: debug
    workingDir: /project
  id: debug
- exec:
    commandLine: npm test
    component: runtime
    group:
      isDefault: true
      kind: test
    workingDir: /project
  id: test
components:
- container:
    endpoints:
    - name: http-3000
      targetPort: 3000
    image: registry.access.redhat.com/ubi8/nodejs-14:latest
    memoryLimit: 1024Mi
    mountSources: true
    sourceMapping: /project
  name: runtime
metadata:
  description: Stack with NodeJS 12
  displayName: NodeJS Runtime
  language: nodejs
  name: nodejs
  projectType: nodejs
  tags:
  - NodeJS
  - Express
  - ubi8
  version: 1.0.0
schemaVersion: 2.0.0
starterProjects:
- git:
    remotes:
      origin: https://github.com/odo-devfiles/nodejs-ex.git
  name: nodejs-starter
MacBook-Pro:node-test stephanie$ odo push

Validation
 ✓  Validating the devfile [38333ns]

Creating Kubernetes resources for component nodejs-node-test-wzrw
 ✓  Waiting for component to start [4s]
 ✓  Waiting for component to start [7ms]
 ⚠  Unable to create ingress, missing host information for Endpoint http-3000, please check instructions on URL creation (refer `odo url create --help`)


Applying URL changes
W0525 11:17:13.050908   89044 warnings.go:70] extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
 ✓  URLs are synced with the cluster, no changes are required.

Syncing to component nodejs-node-test-wzrw
 ✓  Checking files for pushing [2ms]
 ✓  Syncing files to the component [286ms]

Executing devfile commands for component nodejs-node-test-wzrw
 ✓  Waiting for component to start [1ms]
 ✓  Executing install command "npm install" [5s]
 ✓  Executing run command "npm start" [1s]

Pushing devfile component "nodejs-node-test-wzrw"
 ✓  Changes successfully pushed to component
$ odo service list
NAME                                      AGE
ServiceBinding/example-servicebinding     19m26s

```

**Comment by mohammedzee1000** [comment]
Got caught up in release and other stuff and hence still not done investigating

**Comment by mohammedzee1000** [comment]
Ok reproduced:
```bash
> odo create nodejs
Devfile Object Validation
 ✓  Checking devfile existence [75402ns]
 ✓  Creating a devfile component from registry: DefaultDevfileRegistry [110755ns]
Validation
 ✓  Validating if devfile name is correct [100926ns]

Please use `odo push` command to create the component with source deployed
> odo push

Validation
 ✓  Validating the devfile [187000ns]

Creating Kubernetes resources for component nodejs-node-ex-vvuc
 ✓  Waiting for component to start [2s]
 ✓  Waiting for component to start [403ms]

Applying URL changes
W0617 17:12:43.124826   98070 warnings.go:70] extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
 ✓  URL http-3000: http://http-3000-nodejs-node-ex-vvuc-blah2.apps.testocp47.psiodo.net/ created

Syncing to component nodejs-node-ex-vvuc
 ✓  Checking files for pushing [10ms]
 ✓  Syncing files to the component [9s]

Executing devfile commands for component nodejs-node-ex-vvuc
 ✓  Waiting for component to start [446ms]
 ✓  Executing install command "npm install" [19s]
 ✓  Executing run command "npm start" [3s]

Pushing devfile component "nodejs-node-ex-vvuc"
 ✓  Changes successfully pushed to component
> odo catalog list services
Services available through Operators
NAME                                CRDs
etcdoperator.v0.9.4-clusterwide     EtcdCluster, EtcdBackup, EtcdRestore
service-binding-operator.v0.7.1     ServiceBinding
> odo service list
 ✗  no operator backed services found in namespace: blah2
> odo service create etcdoperator.v0.9.4-clusterwide/EtcdCluster
Successfully added service to the configuration; do 'odo push' to create service on the cluster
> odo push
 ✓  Waiting for component to start [389ms]

Validation
 ✓  Validating the devfile [156399ns]

Creating Kubernetes resources for component nodejs-node-ex-vvuc
 ✓  Created service "EtcdCluster/example" on the cluster; refer "odo link -h" to know how to link it to the component
 ✓  Waiting for component to start [371ms]
 ✓  Waiting for component to start [372ms]

Applying URL changes
W0617 17:14:31.470370   98288 warnings.go:70] extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
 ✓  URLs are synced with the cluster, no changes are required.

Syncing to component nodejs-node-ex-vvuc
 ✓  Checking file changes for pushing [15ms]
 ✓  Syncing files to the component [13s]

Executing devfile commands for component nodejs-node-ex-vvuc
 ✓  Waiting for component to start [763ms]
 ✓  Executing install command "npm install" [4s]
 ✓  Executing run command "npm start" [1m]

Pushing devfile component "nodejs-node-ex-vvuc"
 ✓  Changes successfully pushed to component
> odo service list
NAME                    MANAGED BY ODO                STATE      AGE
EtcdCluster/example     Yes (nodejs-node-ex-vvuc)     Pushed     8m27s
> odo delete
? Are you sure you want to delete the devfile component: nodejs-node-ex-vvuc? Yes

Gathering information for component nodejs-node-ex-vvuc
 ✓  Checking status for component [423ms]

Deleting component nodejs-node-ex-vvuc
 ✓  Deleting Kubernetes resources for component [453ms]
 ✓  Successfully deleted component
> odo service list
NAME                    MANAGED BY ODO                STATE      AGE
EtcdCluster/example     Yes (nodejs-node-ex-vvuc)     Pushed     8m58s
> oc get pods
NAME                 READY     STATUS    RESTARTS   AGE
example-jvqthb6d62   1/1       Running   0          4m49s
example-kkgpc4h2ck   1/1       Running   0          5m22s
example-l4ff8cwb6p   1/1       Running   0          5m5s
> odo service delete EtcdCluster/example
? Are you sure you want to delete EtcdCluster/example Yes
Service "EtcdCluster/example" has been successfully deleted; do 'odo push' to delete service from the cluster
> odo push

Validation
 ✓  Validating the devfile [93718ns]

Creating Kubernetes resources for component nodejs-node-ex-vvuc
 ✓  Deleted service "EtcdCluster/example" from the cluster
 ✓  Waiting for component to start [2s]
 ✓  Waiting for component to start [411ms]

Applying URL changes
W0617 17:20:37.786397   99425 warnings.go:70] extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
 ✓  URL http-3000: http://http-3000-nodejs-node-ex-vvuc-blah2.apps.testocp47.psiodo.net/ created

Syncing to component nodejs-node-ex-vvuc
 ✓  Checking files for pushing [11ms]
 ✓  Syncing files to the component [9s]

Executing devfile commands for component nodejs-node-ex-vvuc
 ✓  Waiting for component to start [413ms]
 ✓  Executing install command "npm install" [21s]
 ✓  Executing run command "npm start" [4s]

Pushing devfile component "nodejs-node-ex-vvuc"
 ✓  Changes successfully pushed to component
> odo service list
 ✗  no operator backed services found in namespace: blah2

```
So I can conclude that:
 - push does not delete service by itself
 - workaround is doing `odo service delete blah/blah` before doing `odo delete` and `odo push`
 - Basically we need ot list attached services and delete them as part of push

Further investigation is needed to get more specific

**Comment by mohammedzee1000** [comment]
So I can conclude that:
 - push does not delete service by itself
 - doing `odo service delete blah/blah` before doing `odo delete` followed by `odo push` successfully deletes the service
 - Basically, we need to list attached services and delete them as part of a push. But do we want to do it or not ?

Further investigation is needed to get more specific

**Comment by mohammedzee1000** [comment]
Looking into if Owners Reference can help solve this problem
If it does not work out, we might have to do what service delete does

**Comment by mohammedzee1000** [comment]
After some investigation, it was found that this problem is being fixed by #4819 
cc @dharmit 

```bash
> odo service list
 ✗  no operator backed services found in namespace: st
> odo catalog list services
no deployable services/operators found
> odo create nodejs
Devfile Object Validation
 ✓  Checking devfile existence [107578ns]
 ✓  Creating a devfile component from registry: DefaultDevfileRegistry [196790ns]
Validation
 ✓  Validating if devfile name is correct [152807ns]

Please use `odo push` command to create the component with source deployed
> odo push

Validation
 ✓  Validating the devfile [183747ns]

Updating services

Creating Kubernetes resources for component nodejs-node-ex-nmqy
 ✓  Waiting for component to start [6s]
 ✓  Waiting for component to start [383ms]

Applying URL changes
W0623 15:53:38.596322  281674 warnings.go:70] extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
 ✓  URL http-3000: http://http-3000-nodejs-node-ex-nmqy-st.apps.testocp47.psiodo.net/ created

Syncing to component nodejs-node-ex-nmqy
 ✓  Checking files for pushing [16ms]
 ✓  Syncing files to the component [9s]

Executing devfile commands for component nodejs-node-ex-nmqy
 ✓  Waiting for component to start [369ms]
 ✓  Executing install command "npm install" [59s]
 ✓  Executing run command "npm start" [3s]

Pushing devfile component "nodejs-node-ex-nmqy"
 ✓  Changes successfully pushed to component
 > odo catalog list services
Services available through Operators
NAME                                CRDs
etcdoperator.v0.9.4-clusterwide     EtcdCluster, EtcdBackup, EtcdRestore
service-binding-operator.v0.7.1     ServiceBinding
> odo service list
 ✗  no operator backed services found in namespace: st
> odo service create etcdoperator.v0.9.4-clusterwide/EtcdCluster
Successfully added service to the configuration; do 'odo push' to create service on the cluster
> odo service list
NAME                    MANAGED BY ODO                STATE          AGE
EtcdCluster/example     Yes (nodejs-node-ex-nmqy)     Not pushed     
> odo push
 ✓  Waiting for component to start [421ms]

Validation
 ✓  Validating the devfile [348057ns]

Updating services
 ✓  Created service "EtcdCluster/example" on the cluster; refer "odo link -h" to know how to link it to the component

Creating Kubernetes resources for component nodejs-node-ex-nmqy
 ✓  Waiting for component to start [366ms]
 ✓  Waiting for component to start [377ms]

Applying URL changes
W0623 15:56:45.439086  282080 warnings.go:70] extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
 ✓  URLs are synced with the cluster, no changes are required.

Syncing to component nodejs-node-ex-nmqy
 ✓  Checking file changes for pushing [15ms]
 ✓  Syncing files to the component [6s]

Executing devfile commands for component nodejs-node-ex-nmqy
 ✓  Waiting for component to start [379ms]
 ✓  Executing install command "npm install" [4s]
 ✓  Executing run command "npm start" [4s]

Pushing devfile component "nodejs-node-ex-nmqy"
 ✓  Changes successfully pushed to component
> odo service list
NAME                    MANAGED BY ODO                STATE      AGE
EtcdCluster/example     Yes (nodejs-node-ex-nmqy)     Pushed     5m0s
> kubectl get pods
NAME                                       READY     STATUS    RESTARTS   AGE
example-5t476nc5w9                         1/1       Running   0          <invalid>
example-bdb6js2lfl                         1/1       Running   0          <invalid>
example-rdwg45ftj7                         1/1       Running   0          <invalid>
nodejs-node-ex-nmqy-app-6d79c7769b-jch2c   1/1       Running   0          8m6s
> odo link EtcdCluster/example --name etcd
 ✓  Successfully created link between component "nodejs-node-ex-nmqy" and service "EtcdCluster/example"

To apply the link, please use `odo push`
> odo push
 ✓  Waiting for component to start [396ms]

Validation
 ✓  Validating the devfile [189053ns]

Updating services
 ✓  Created link "etcd" on the cluster; component will be restarted
 ✓  Waiting for component to start [8s]

Creating Kubernetes resources for component nodejs-node-ex-nmqy
 ✓  Waiting for component to start [12s]
 ✓  Waiting for component to start [515ms]
W0623 15:58:57.543145  282248 warnings.go:70] .metadata.ownerReferences contains duplicate entries; API server dedups owner references in 1.20+, and may reject such requests as early as 1.24; please fix your requests; duplicate UID(s) observed: e7ae1339-6a40-4891-8dcc-f9131a5a6ffd

Applying URL changes
W0623 15:58:59.799231  282248 warnings.go:70] extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
 ✓  URLs are synced with the cluster, no changes are required.

Syncing to component nodejs-node-ex-nmqy
 ✓  Checking file changes for pushing [15ms]
 ✓  Syncing files to the component [9s]

Executing devfile commands for component nodejs-node-ex-nmqy
 ✓  Waiting for component to start [370ms]
 ✓  Executing install command "npm install" [13s]
 ✓  Executing run command "npm start" [4s]

Pushing devfile component "nodejs-node-ex-nmqy"
 ✓  Changes successfully pushed to component
> odo service list
NAME                    MANAGED BY ODO                STATE      AGE
EtcdCluster/example     Yes (nodejs-node-ex-nmqy)     Pushed     9m21s
ServiceBinding/etcd     Yes (nodejs-node-ex-nmqy)     Pushed     9m19s
> kubectl get pods
NAME                                       READY     STATUS    RESTARTS   AGE
example-6x45vsllc2                         1/1       Running   0          83s
example-ftd2jsq6dk                         1/1       Running   0          64s
example-j7tq6czc6v                         1/1       Running   0          99s
nodejs-node-ex-nmqy-app-678f6b9d57-r72zj   1/1       Running   0          9m50s
> odo delete
? Are you sure you want to delete the devfile component: nodejs-node-ex-nmqy? Yes

Gathering information for component nodejs-node-ex-nmqy
 ✓  Checking status for component [762ms]

Deleting component nodejs-node-ex-nmqy
 ✓  Deleting Kubernetes resources for component [402ms]
 ✓  Successfully deleted component
> odo service list
NAME                    MANAGED BY ODO                STATE          AGE
EtcdCluster/example     Yes (nodejs-node-ex-nmqy)     Not pushed     
ServiceBinding/etcd     Yes (nodejs-node-ex-nmqy)     Not pushed     
> kubectl get pods
No resources found.
> odo service delete EtcdCluster/example
? Are you sure you want to delete EtcdCluster/example Yes
Service "EtcdCluster/example" has been successfully deleted; do 'odo push' to delete service from the cluster
> odo service list
NAME                    MANAGED BY ODO                STATE          AGE
ServiceBinding/etcd     Yes (nodejs-node-ex-nmqy)     Not pushed     

```
